doc/configuration.rst


Configuration
=============

The default configuration for the bundle looks like this:

.. code-block:: yaml

    # app/config/config.yml
    liip_imagine:

        resolvers:
            default:
                web_path:
                    web_root: ~ # %kernel.project_dir%/public (%kernel.project_dir%/web for Symfony < 4.0.0)
                    cache_prefix: ~ # media/cache

        loaders:
            default:
                filesystem:
                    data_root: ~  # %kernel.project_dir%/public (%kernel.project_dir%/web for Symfony < 4.0.0)

        driver:               gd
        cache:                default
        data_loader:          default
        default_image:        null
        twig:
            mode:             legacy
        default_filter_set_settings:
            quality:              100
            jpeg_quality:         ~
            png_compression_level:  ~
            png_compression_filter: ~
            animated:             false
            format:               ~
            cache:                ~
            data_loader:          ~
            default_image:        null
            filters: []
            post_processors: []
        controller:
            filter_action:          liip_imagine.controller::filterAction
            filter_runtime_action:  liip_imagine.controller::filterRuntimeAction
            redirect_response_code: 302
        webp:
            generate:    false
            quality:     100
            cache:       ~
            data_loader: ~
            post_processors: []
        filter_sets:

            # Prototype
            name:
                quality:              100
                jpeg_quality:         ~
                png_compression_level:  ~
                png_compression_filter: ~
                animated:             false
                format:               ~
                cache:                ~
                data_loader:          ~
                default_image:        null
                filters:

                    # Prototype
                    name:                 []

                post_processors:

                    # Prototype
                    name:                 []

There are several configuration options available:

* ``cache`` - default cache resolver. Default value: ``web_path`` (which means
  the standard web_path resolver is used)
* ``data_loader`` - name of a custom data loader. Default value: ``filesystem``
  (which means the standard filesystem loader is used). Built-in loaders include:
  ``filesystem``, ``chain``, ``flysystem``, ``stream``, and ``asset_mapper``.
* ``twig.mode`` - Twig filter integration. ``none`` disables the twig filters, ``lazy`` enables
  Twig using the Twig runtime for lazy loading. The default value is ``lazy``.
  The twig filter automatically picks up the ``framework.assets.version`` configuration. You can
  overwrite the version with the ``twig.assets_version`` option. See :doc:`asset-versioning` for
  more information.
* ``controller``
    * ``filter_action`` - name of the controller action to use in the route loader.
      Default value: ``liip_imagine.controller:filterAction``
    * ``filter_runtime_action`` - name of the controller action to use in the route
      loader for runtimeconfig images. Default value: ``liip_imagine.controller:filterRuntimeAction``
    * ``redirect_response_code`` - The HTTP redirect response code to return from the imagine controller,
      one of ``201``, ``301``, ``302``, ``303``, ``307``, or ``308``. Default value: ``302``
      See :doc:`optimizations/avoid-redirects` if you want to change this configuration.
* ``webp``
    * ``generate`` - enabling the generation a copy of the image in the WebP format.
    * ``quality`` - override the quality from filter option.
    * ``cache`` - default cache resolver. Default value: ``web_path`` (which means
      the standard web_path resolver is used)
    * ``data_loader`` - name of a custom data loader. Default value: ``filesystem``
      (which means the standard filesystem loader is used). Built-in loaders include:
      ``filesystem``, ``chain``, ``flysystem``, ``stream``, and ``asset_mapper``.
    * ``post_processors`` - sets post-processors to be applied on filtered image
      (see Post-Processors section in the :doc:`filters chapter <filters>` for details).
* ``driver`` - one of the drivers: ``gd``, ``imagick``, ``gmagick``, ``vips``.
  Default value: ``gd``
  * If you want to use vips, you need to additionally require ``rokka/imagine-vips``
* ``default_filter_set_settings`` - specify the default values that will be inherit for any set defined in
  ``filter_sets``. These values will be overridden if they are specified in the each set. In case of ``filters`` and
  ``post_processors``, the specified values will be merged with the default ones.
* ``filter_sets`` - specify the filter sets that you want to define and use.

Each filter set that you specify has the following options:

* ``filters`` - determine the type of filter to be used (refer to *Filters* section
  for more information) and options that should be passed to the specific filter type.
* ``post_processors`` - sets post-processors to be applied on filtered image
  (see Post-Processors section in the :doc:`filters chapter <filters>` for details).
* ``jpeg_quality`` - override the quality for jpeg images (this overrides the
  ``quality`` option above)
* ``png_compression_level`` - set the compression level for png images (0-9)
  (this overrides the ``quality`` option above)
* ``png_compression_filter`` - set the compression filter for png images (see the
  ``filters`` parameter for ``imagepng`` function in `PHP manual`_ for more details)
* ``cache`` - override the default cache setting.
* ``data_loader`` - override the default data loader.
* ``route`` - optional list of route requirements, defaults and options using in
  the route loader. Add array with keys ``requirements``, ``defaults`` or ``options``.
  Default value: empty array.
* ``format`` - hardcodes the output format (which means that the requested format
  is ignored).
* ``animated`` - support for resizing animated gif (currently not supported by
  Imagine (PR pending))


.. _`PHP Manual`: https://php.net/imagepng


Resources/doc/configuration.md
# Configuration

The default configuration for the bundle looks like this:

``` yaml
liip_imagine:

    resolvers:
        default:
            web_path:
              web_root: ~ # %kernel.root_dir%/../web
              cache_prefix: ~ # media/cache


    loaders:
        default:
            filesystem:
                data_root: ~  # %kernel.root_dir%/../web/

    driver:               gd
    cache:                default
    data_loader:          default
    default_image:        null
    controller:
        filter_action:         liip_imagine.controller:filterAction
        filter_runtime_action: liip_imagine.controller:filterRuntimeAction
    filter_sets:

        # Prototype
        name:
            quality:              100
            animated:             false
            format:               ~
            cache:                ~
            data_loader:          ~
            default_image:        null
            filters:

                # Prototype
                name:                 []

            post_processors:

                # Prototype
                name:                 []
```

There are several configuration options available:

 - `cache` - default cache resolver

    default: web_path (which means the standard web_path resolver is used)

 - `data_loader` - name of a custom data loader

    default: filesystem (which means the standard filesystem loader is used)

 - `controller`
         - `filter_action` - name of the controller action to use in the route loader

            default: liip_imagine.controller:filterAction

        - `filter_runtime_action` - name of the controller action to use in the route loader for runtimeconfig images

            default: liip_imagine.controller:filterRuntimeAction

 - `driver` - one of the three drivers: `gd`, `imagick`, `gmagick`

    default: `gd`

 - `filter_sets` - specify the filter sets that you want to define and use

Each filter set that you specify has the following options:

 - `filters` - determine the type of filter to be used (refer to *Filters* section for more information)
    and options that should be passed to the specific filter type
 - `post_processors` - sets post-processors to be applied on filtered image (see [Post-Processors](filters.md#post-processors) for details)
 - `quality` - override the default quality of 100 for the generated images
 - `cache` - override the default cache setting
 - `data_loader` - override the default data loader
 - `route` - optional list of route requirements, defaults and options using in the route loader. Add array with keys 'requirements', 'defaults' or 'options'.

    default: empty array

 - `format` - hardcodes the output format (aka the requested format is ignored)
 - `animated` - support for resizing animated gif (currently not supported by Imagine (PR pending))

[Back to the index](index.md)


CHANGELOG.md
# Changelog

This file contains a complete enumeration of all [pull requests](https://github.com/liip/LiipImagineBundle/pulls) merged
for a given releases. Unreleased, upcoming changes will be updated here periodically; reference the next release on our
[milestones](https://github.com/liip/LiipImagineBundle/milestones) page for the latest changes.

## 3.x (unreleased)

- Dropped support for PHP templating [#1549](https://github.com/liip/LiipImagineBundle/pull/1549).
- Dropped support for old PHP versions, minimum version is now 8.1 [#1549](https://github.com/liip/LiipImagineBundle/pull/1549).
- Removed legacy `AmazonS3Resolver` for the outdated `amazonwebservices/aws-sdk-for-php`, use `AwsS3Resolver` with `aws/aws-sdk-php` instead [#1549](https://github.com/liip/LiipImagineBundle/pull/1549).
- Removed deprecated code and BC for old Symfony versions.
- The response when the `filter` parameter in a resolve request is not an array is now 400 bad request, and no longer 404 not found.
- Simplified the `ProxyResolver` to no longer do the undocumented regex replace logic on domain names (reverting [#687](https://github.com/liip/LiipImagineBundle/pull/687)).
- Removed `use_psr_cache`, the cache now always needs to be a PSR cache and Doctrine cache is no longer supported.

# 2.x

## [2.17.1](https://github.com/liip/LiipImagineBundle/tree/2.17.1)

- Fix for AssetMapperLoaderFactory service definition to not fail when not configured ([dbu](https://github.com/liip/LiipImagineBundle/pull/1649))

## [2.17.0](https://github.com/liip/LiipImagineBundle/tree/2.17.0)

- Add AssetMapperLocator to work with the asset mapper in development mode ([tito10047](https://github.com/liip/LiipImagineBundle/pull/1645))
- Drop support for PHP 7 ([dbu](https://github.com/liip/LiipImagineBundle/pull/1646))

## [2.16.0](https://github.com/liip/LiipImagineBundle/tree/2.16.0)

- Compatible with Symfony 8 ([dmaicher](https://github.com/liip/LiipImagineBundle/pull/1642)) (including a bunch of PRs to change all XML configuration to PHP configuration)
- Test with PHP 8.5 and fix deprecations([dbu](https://github.com/liip/LiipImagineBundle/pull/1644))
- Drop support for unmaintained Symfony versions ([dmaicher](https://github.com/liip/LiipImagineBundle/pull/1639))

## [2.15.0](https://github.com/liip/LiipImagineBundle/tree/2.15.0)

- Refactored `Request::get()` to `Request::query::get()` to avoid deprecation with Symfony 7.4 ([dmaicher](https://github.com/liip/LiipImagineBundle/pull/1636))

## [2.14.0](https://github.com/liip/LiipImagineBundle/tree/2.14.0)

- Allow configuring empty ACLs on AWS S3 resolver to skip ACLs ([deguif](https://github.com/liip/LiipImagineBundle/pull/1629))
- Add `use_psr_cache` option to AWS S3 resolver configuration in order to support PSR cache. Configuring a doctrine cache now will trigger a deprecation ([deguif](https://github.com/liip/LiipImagineBundle/pull/1630))
- Add ability to specify a custom s3 client service in aws s3 resolver configuration ([deguif](https://github.com/liip/LiipImagineBundle/pull/1632))

## [2.13.3](https://github.com/liip/LiipImagineBundle/tree/2.13.3)

- Prevent InvalidArgumentException from FileinfoMimeTypeGuesser when chain loading an image that is not a file ([revoltek-daniel](https://github.com/liip/LiipImagineBundle/pull/1614))

## [2.13.2](https://github.com/liip/LiipImagineBundle/tree/2.13.2)

- Remove deprecated spaceless filter from twig template ([JohJohan](https://github.com/liip/LiipImagineBundle/pull/1609))

## [2.13.1](https://github.com/liip/LiipImagineBundle/tree/2.13.1)

- Fix Json Manifest handling when manifest file does not exist yet ([AirBair](https://github.com/liip/LiipImagineBundle/pull/1600))

## [2.13.0](https://github.com/liip/LiipImagineBundle/tree/2.13.0)

- Support JsonManifestVersionStrategy that was added in Symfony 6 ([wouterSkepp](https://github.com/liip/LiipImagineBundle/pull/1529)).

## [2.12.3](https://github.com/liip/LiipImagineBundle/tree/2.12.3)

- Add alias for `Imagine\Image\ImagineInterface` to help autowiring ([dbu](https://github.com/liip/LiipImagineBundle/pull/1583)). 

## [2.12.2](https://github.com/liip/LiipImagineBundle/tree/2.12.2)

- Autoconfigure tags `'liip_imagine.filter.loader`, `liip_imagine.filter.post_processor` and `liip_imagine.binary.loader`
  and allow to specify these by class name in addition to service names ([homersimpsons](https://github.com/liip/LiipImagineBundle/pull/1486))
- Avoid PHP 8.3 warning with default empty prefix ([simonberger](https://github.com/liip/LiipImagineBundle/pull/1568))

## [2.12.1](https://github.com/liip/LiipImagineBundle/tree/2.12.1)

- Adjustments to install with Symfony 7 ([mbabker](https://github.com/liip/LiipImagineBundle/pull/1535))

## [2.12.0](https://github.com/liip/LiipImagineBundle/tree/2.12.0)

- Fix documentation filter command parameter name ([rdavaillaud](https://github.com/liip/LiipImagineBundle/pull/1515))
- Add return types in form ([garak](https://github.com/liip/LiipImagineBundle/pull/1518)) (fixes support of new Symfony versions)
- Improved documentation for flysystem adapter ([gregberger](https://github.com/liip/LiipImagineBundle/pull/1522))
- Drop support for PHP 7.1, minimum version is now 7.2.
- PHPStan and CS fixes

## [2.11.0](https://github.com/liip/LiipImagineBundle/tree/2.11.0)

- Compatibility with Symfony 6.3 (We do not expect users to extend a compiler passes or the DI extension of this bundle. If you did, you might need to adjust return types) ([mbabker](https://github.com/liip/LiipImagineBundle/pull/1514))
- Documentation improvements ([StevenRenaux](https://github.com/liip/LiipImagineBundle/pull/1511), [StevenRenaux](https://github.com/liip/LiipImagineBundle/pull/1510), [mysterty](https://github.com/liip/LiipImagineBundle/pull/1507))

## [2.10.0](https://github.com/liip/LiipImagineBundle/tree/2.10.0)


src/DependencyInjection/Configuration.php
<?php

/*
 * This file is part of the `liip/LiipImagineBundle` project.
 *
 * (c) https://github.com/liip/LiipImagineBundle/graphs/contributors
 *
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 */

namespace Liip\ImagineBundle\DependencyInjection;

use Liip\ImagineBundle\Config\Controller\ControllerConfig;
use Liip\ImagineBundle\Controller\ImagineController;
use Liip\ImagineBundle\DependencyInjection\Factory\FactoryInterface;
use Liip\ImagineBundle\DependencyInjection\Factory\Loader\LoaderFactoryInterface;
use Liip\ImagineBundle\DependencyInjection\Factory\Resolver\ResolverFactoryInterface;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;
use Symfony\Component\Config\Definition\Exception\InvalidConfigurationException;

class Configuration implements ConfigurationInterface
{
    /**
     * @var ResolverFactoryInterface[]
     */
    protected array $resolversFactories;

    /**
     * @var LoaderFactoryInterface[]
     */
    protected array $loadersFactories;

    /**
     * @param ResolverFactoryInterface[] $resolversFactories
     * @param LoaderFactoryInterface[]   $loadersFactories
     */
    public function __construct(array $resolversFactories, array $loadersFactories)
    {
        $this->resolversFactories = $resolversFactories;
        $this->loadersFactories = $loadersFactories;
    }

    public function getConfigTreeBuilder(): TreeBuilder
    {
        $treeBuilder = new TreeBuilder('liip_imagine');
        $rootNode = $treeBuilder->getRootNode();

        $resolversPrototypeNode = $rootNode
            ->children()
                ->arrayNode('resolvers')
                    ->useAttributeAsKey('name')
                    ->prototype('array')
                        ->performNoDeepMerging();
        $this->addResolversSections($resolversPrototypeNode);

        $loadersPrototypeNode = $rootNode
            ->children()
                ->arrayNode('loaders')
                    ->useAttributeAsKey('name')
                    ->prototype('array');
        $this->addLoadersSections($loadersPrototypeNode);

        $rootNode
            ->beforeNormalization()
                ->ifTrue(static function ($v) {
                    return
                        empty($v['loaders'])
                        || empty($v['loaders']['default'])
                        || empty($v['resolvers'])
                        || empty($v['resolvers']['default']);
                })
                ->then(static function ($v) {
                    if (empty($v['loaders'])) {
                        $v['loaders'] = [];
                    }

                    if (false === \is_array($v['loaders'])) {
                        throw new \LogicException('Loaders has to be array');
                    }

                    if (false === \array_key_exists('default', $v['loaders'])) {
                        $v['loaders']['default'] = ['filesystem' => null];
                    }

                    if (empty($v['resolvers'])) {
                        $v['resolvers'] = [];
                    }

                    if (false === \is_array($v['resolvers'])) {
                        throw new \LogicException('Resolvers has to be array');
                    }

                    if (false === \array_key_exists('default', $v['resolvers'])) {
                        $v['resolvers']['default'] = ['web_path' => null];
                    }

                    return $v;
                })
            ->end();

        $rootNode
            ->fixXmlConfig('filter_set', 'filter_sets')
            ->children()
                ->scalarNode('driver')->defaultValue('gd')
                    ->validate()
                        ->ifTrue(static function ($v) {
                            return !\in_array($v, ['gd', 'imagick', 'gmagick', 'vips'], true);
                        })
                        ->thenInvalid('Invalid imagine driver specified: %s')
                    ->end()
                ->end()
                ->scalarNode('cache')->defaultValue('default')->end()
                ->scalarNode('cache_base_path')->defaultValue('')->end()
                ->scalarNode('data_loader')->defaultValue('default')->end()
                ->scalarNode('default_image')->defaultNull()->end()
                ->arrayNode('default_filter_set_settings')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('quality')->defaultValue(100)->end()
                        ->scalarNode('jpeg_quality')->defaultNull()->end()
                        ->scalarNode('png_compression_level')->defaultNull()->end()
                        ->scalarNode('png_compression_filter')->defaultNull()->end()
                        ->scalarNode('format')->defaultNull()->end()
                        ->booleanNode('animated')->defaultFalse()->end()
                        ->scalarNode('cache')->defaultNull()->end()
                        ->scalarNode('data_loader')->defaultNull()->end()
                        ->scalarNode('default_image')->defaultNull()->end()
                        ->arrayNode('filters')
                            ->useAttributeAsKey('name')
                            ->prototype('array')
                                ->useAttributeAsKey('name')
                                ->ignoreExtraKeys(false)
                                ->prototype('variable')->end()
                            ->end()
                        ->end()
                        ->arrayNode('post_processors')
                            ->defaultValue([])
                            ->useAttributeAsKey('name')
                            ->prototype('array')
                                ->useAttributeAsKey('name')
                                ->ignoreExtraKeys(false)
                                ->prototype('variable')->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
                ->arrayNode('controller')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('filter_action')->defaultValue(\sprintf('%s::filterAction', ImagineController::class))->end()
                        ->scalarNode('filter_runtime_action')->defaultValue(\sprintf('%s::filterRuntimeAction', ImagineController::class))->end()
                        ->integerNode('redirect_response_code')->defaultValue(302)
                            ->validate()
                                ->ifTrue(static function ($redirectResponseCode) {
                                    return !\in_array($redirectResponseCode, ControllerConfig::REDIRECT_RESPONSE_CODES, true);
                                })
                                ->thenInvalid('Invalid redirect response code "%s" (must be 201, 301, 302, 303, 307, or 308).')
                            ->end()
                        ->end()
                    ->end()
                ->end()
                ->arrayNode('filter_sets')
                    ->useAttributeAsKey('name')
                    ->prototype('array')
                        ->fixXmlConfig('filter', 'filters')
                        ->children()
                            ->scalarNode('quality')->end()
                            ->scalarNode('jpeg_quality')->end()
                            ->scalarNode('png_compression_level')->end()
                            ->scalarNode('png_compression_filter')->end()
                            ->scalarNode('format')->end()
                            ->booleanNode('animated')->end()
                            ->scalarNode('cache')->end()
                            ->scalarNode('data_loader')->end()
                            ->scalarNode('default_image')->end()
                            ->arrayNode('filters')
                                ->useAttributeAsKey('name')
                                ->prototype('array')
                                    ->useAttributeAsKey('name')
                                    ->ignoreExtraKeys(false)
                                    ->prototype('variable')->end()
                                ->end()
                            ->end()
                            ->arrayNode('post_processors')
                                ->defaultValue([])
                                ->useAttributeAsKey('name')
                                ->prototype('array')
                                    ->useAttributeAsKey('name')
                                    ->ignoreExtraKeys(false)
                                    ->prototype('variable')->end()
                                ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
            ->arrayNode('twig')
                ->addDefaultsIfNotSet()
                ->children()
                    ->enumNode('mode')
                        ->defaultValue('lazy')
                        ->info('Twig mode: none/lazy (default)')
                        ->values(['none', 'lazy'])
                    ->end()
                    ->scalarNode('assets_version')->defaultNull()->end()
                ->end()
            ->end()
            ->arrayNode('messenger')
                ->canBeEnabled()
                ->info('Enables integration with symfony/messenger if set true. Warmup image caches in background by sending messages to MQ.')
            ->end()
        ->end();

        $rootNode
            ->children()
                ->arrayNode('webp')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->booleanNode('generate')->defaultFalse()->end()
                        ->integerNode('quality')->defaultValue(100)->end()
                        ->scalarNode('cache')->defaultNull()->end()
                        ->scalarNode('data_loader')->defaultNull()->end()
                        ->arrayNode('post_processors')
                            ->defaultValue([])
                            ->useAttributeAsKey('name')
                            ->prototype('array')
                                ->useAttributeAsKey('name')
                                ->ignoreExtraKeys(false)
                                ->prototype('variable')->end()
                            ->end()
                    ->end()
                ->end()
            ->end();

        return $treeBuilder;
    }

    private function addResolversSections(ArrayNodeDefinition $resolversPrototypeNode): void
    {
        $this->addConfigurationSections($this->resolversFactories, $resolversPrototypeNode, 'resolver');
    }

    private function addLoadersSections(ArrayNodeDefinition $resolversPrototypeNode): void
    {
        $this->addConfigurationSections($this->loadersFactories, $resolversPrototypeNode, 'loader');
    }

    /**
     * @param FactoryInterface[] $factories
     */
    private function addConfigurationSections(array $factories, ArrayNodeDefinition $definition, string $type): void
    {
        foreach ($factories as $f) {
            $f->addConfiguration($definition->children()->arrayNode($f->getName()));
        }

        $nodeDefinition = $definition->end();

        \assert($nodeDefinition instanceof ArrayNodeDefinition);

        $nodeDefinition
            ->validate()
            ->ifTrue(static function ($array) use ($type) {
                foreach ($array as $name => $element) {
                    if (!$element) {
                        throw new InvalidConfigurationException(ucfirst($type).' "'.$name.'" must have a factory configured');
                    }
                }

                return false;
            })
            ->thenInvalid('Each '.$type.' must have a factory configured')
            ->end();
    }
}
